### PR TITLE
AVRO-3486: Set namespace with protocol fullName

### PR DESCRIPTION
**Pull request by opwvhk**
When parsing or creating a protocol, use both the name and namespace
given to determine the actual name, namespace and fullName.

Reason: the spec says "The name and namespace qualification rules
defined for schema objects apply to protocols as well."

Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3486
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [X] My PR ~adds~ updates the following unit tests __OR__ does not need testing for this extremely good reason:
    `org.apache.avro.compiler.idl.TestIdl#runTests()` for `import.avdl` and `nestedimport.avdl` by updating `bar.avpr`

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: lang/java/avro/src/main/java/org/apache/avro/Protocol.java [hunk 1]
```diff
@@ -275,15 +274,29 @@ public Protocol(Protocol p) {
...
   public Protocol(String name, String namespace) {
     this(name, null, namespace);
   }
 
+  private void setName(String name, String namespace) {
+    int lastDot = name.lastIndexOf('.');
+    if (lastDot < 0) {
+      this.name = name;
+      this.namespace = namespace;
+    } else {
```
Comment (Christophe-LeSaec) [inline lang/java/avro/src/main/java/org/apache/avro/Protocol.java:290]: What if namespace is "java" and name is "lang.String" 
Here, a suggestion :
```java
        StringBuilder namespaceBuilder = new StringBuilder((namespace != null) ? namespace : "");
        if (namespaceBuilder.length() > 0 && namespaceBuilder.charAt(namespaceBuilder.length() - 1) != '.') {
            namespaceBuilder.append('.');
        }
        namespaceBuilder.append(name.substring(0, lastDot));
       this.namespace = namespaceBuilder.toString();
```

File: lang/java/avro/src/main/java/org/apache/avro/Protocol.java [hunk 2]
```diff
@@ -275,15 +274,29 @@ public Protocol(Protocol p) {
...
+      this.name = name.substring(lastDot) + 1;
+      this.namespace = name.substring(0, lastDot);
+    }
+    if ("".equals(this.namespace)) {
```
Comment (Christophe-LeSaec) [inline lang/java/avro/src/main/java/org/apache/avro/Protocol.java:294]: So
```java
if (this.namespace != null && this.namespace.isEmpty()) {
```

**Review by Christophe-LeSaec — COMMENTED** [review]
Good code, 
A proposition for code for very specific corner case;
I also suggest to add unit test directly on [TestProtocol class](https://github.com/apache/avro/blob/master/lang/java/avro/src/test/java/org/apache/avro/TestProtocol.java) (it's more direct than to pass by avro file).

File: lang/java/avro/src/main/java/org/apache/avro/Protocol.java (hunk 1 above)
Comment (opwvhk) [reply lang/java/avro/src/main/java/org/apache/avro/Protocol.java:290]: The intent is to use the same namespace the the definition itself, and any enclosed definitions that don't specify one themselves. This is a lot easier to comprehend than using different namespaces.

File: lang/java/avro/src/main/java/org/apache/avro/Protocol.java (hunk 2 above)
Comment (opwvhk) [reply lang/java/avro/src/main/java/org/apache/avro/Protocol.java]: Changed.
